### DISCUSSION TITLE
Setup reverse proxy in front of input harbor

### DISCUSSION THREAD

**Discussion by jooola** [savonet/liquidsoap#2472] [Ideas]
I am trying to send a stream to a liquidsoap input harbor, behind a reverse proxy. I wonder if this is doable, or if I should use `input.harbor.ssl*` attributes ?

I prefer to have all my TLS certificate on the reverse proxy and terminate the encrypted connection there, instead of managing multiple certs on multiple hosts.

I found a similar demand, but without any answer https://www.mail-archive.com/savonet-users@lists.sourceforge.net/msg14640.html

Basically I try to do the following:
```
BUTT => haproxy on :443 (drop encryption) => liquidsoap on :9901
```

I have also tried the following and it is working:
```
BUTT => liquidsoap on :9901
```

I don't know if BUTT is the right tool to test, but it will be the tool used to stream content.

I would really appreciate some pointers for this!

**Comment by toots** [comment]
Yes, this should work. I hadn't looked at this in a while but it looks like `input.harbor` supports both `PUT` and `POST` methods for sending data to a source mountpoint so you should be able to use a regular HTTP proxy relay. The only question being how long can the proxy send data through a given connection. These are usually designed with finite connections in mind.

Please let us know if you experience any trouble implementing this, this is definitely a use-case that we do want to support.

**Reply by jooola** [reply]
I tried to make this work, but BUTT could not connect on the endpoint using the reverse proxy, an it didn't yield any useful logs.

**Reply by jooola** [reply]
I'll investigate some more to give your something you can work with.

**Reply by jooola** [reply]
I stopped using butt as this isn't a tool modern enough to actually run with tls.

An error on my side, when using TLS, having the port in the host wasn't caught by the filtering matching rule to forward the request to the right backend. Removing it I could hit the right backend. This was also a problem with BUTT as it requires a port in the configure.

I am facing the expected broken pipe. So I'll read some docs, but I don't think a reverse proxy will ever allow a infinite request.

**Reply by jooola** [reply]
From what I read on the net, not having a timeout in the reverse proxy is a huge security problem. Maybe one can loop and reconnect to keep streaming, but that feels dirty.

I think I'll have to find another way to send a stream securely.

**Reply by jooola** [reply]
I tried using a chunked content request, but liquid soap fails with `Feeding stopped: Ogg_demuxer.Invalid_stream.`

**Reply by jooola** [reply]
```bash
input="/media/jo/helvete/audio/Isfar_Sarabski-Planet/01-08-Mark_Guiliana-G-Man-LLS.flac"

server="REDACTED.org"
port="9901"
secure_port="443"
mount="master"
password="REDACTED"
user="admin"

# Using normal port
[[ "$1" = "normal" ]] && ./ffmpeg -v debug -re -vn -i "$input" \
  -f ogg \
  -content_type 'application/ogg' \
  "icecast://$user:$password@$server:$port/$mount"

# Using secure port
[[ "$1" = "secure" ]] && ./ffmpeg -v debug -re -vn -i "$input" \
  -f ogg \
  -content_type 'application/ogg' \
  "icecast://$user:$password@$server:$secure_port/$mount"

# Using secure port with tls option
[[ "$1" = "secure_tls" ]] && ./ffmpeg -v debug -re -vn -i "$input" \
  -f ogg \
  -content_type 'application/ogg' \
  -tls "true" \
  "icecast://$user:$password@$server:$secure_port/$mount"

# Using no port with tls option
[[ "$1" = "secure_no_port" ]] && ./ffmpeg -v debug -re -vn -i "$input" \
  -f ogg \
  -content_type 'application/ogg' \
  -tls "true" \
  "icecast://$user:$password@$server/$mount"

#  Using https (chunked content)
[[ "$1" = "https" ]] && ./ffmpeg -v debug -re -vn -i "$input" \
  -f ogg \
  -content_type 'application/ogg' \
  -method 'PUT' \
  "https://$user:$password@$server/$mount"
```

**Reply by toots** [reply]
I'm not sure that harbor input supports chunked requests at this time. And, yeah, the timeout is gonna be an issue too. I'd say, if you want to use standard HTTP tools, you'd have to look for HLS. We do not have an equivalent of `input.harbor` with HLS.

At the end of the day, you might want to give `input.harbor.ssl` a try.. Or else, you could write your own reverse proxy using node.

**Reply by jooola** [reply]
It seems chunked content request will be supported by Icecast 2.5, this should solve the problems I have !

https://wiki.xiph.org/Icecast_Server/known_reverse_proxy_restrictions
